Safely handle asserts/assumptions in invalid test states

yperess opened the pull request:
If a developer calls a failed `zassert` or `zassume` in the `after` or `teardown` functions, the test should be considered failed as the binary cannot be considered valid. This stems from the fact that at that point, the test was already marked as passing.

This change was validated using a sample test that verified the following:

native_posix:
- Called `zassert_true(false, NULL)` during a suite's `after` function
- Called `zassert_true(false, NULL)` during a suite's `teardown` function
- Called `zassume_true(false, NULL)` during a suite's `after` function
- Called `zassume_true(false, NULL)` during a suite's `teardown` function
- Called `ztest_test_pass()` during a suite's `after` function
- Called `ztest_test_pass()` during a suite's `teardown` function
unit_testing:
- Called `zassert_true(false, NULL)` during a suite's `after` function
- Called `zassert_true(false, NULL)` during a suite's `teardown` function
- Called `zassume_true(false, NULL)` during a suite's `after` function
- Called `zassume_true(false, NULL)` during a suite's `teardown` function
- Called `ztest_test_pass()` during a suite's `after` function
- Called `ztest_test_pass()` during a suite's `teardown` function

In all cases, the final test result was a failure.

Commit:
ztest: remove init_testing

When KERNEL was not defined (unittest), the call to init_testing was
used to set a longjump target using 'stack_fail'. When triggered,
this was actually causing a segfault, because longjmp is only valid
if going directly up the stack. Since init_testing returned, it was
no longer on the stack. Instead, that logic MUST be inlined.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: move end_report implementation

Move the implementation of end_report so it can be used by both code
paths (with and without KERNEL).

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: unittest: Add before/after phase

Add missing setters for the before/after phase when running without
KERNEL.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: unittest: Fix fail/skip/pass outside of tests

When not running in setup, before, or in the test. Calling fail, skip,
or pass is invalid and should be considered an error. Properly handle
these cases by printing a more detailed error message and bailing the
process.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: move get_friendly_phase_name to common code

Move the function used for printing the phase name up so its available
for both unittest and KERNEL mode of tests.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
ztest: Safely handle fail/skip/pass outside tests

Updates the implementation when KERNEL is available to safely bail on
the test when the test calls fail, skip, or pass during invalid test
phases. Print a detailed message, and skip all other tests. The test
run will be marked as failed.

Signed-off-by: Yuval Peress <peress@google.com>

asemjonovs commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -357,6 +388,7 @@ static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test
 	run_test_functions(suite, test, data);
 out:
 	ret |= cleanup_test(test);
```
Can we move `cleanup_test` after calling `after` rules so we have same sequence of events as `run_test` in `KERNEL` enabled builds?

yperess replied:
Done

Commit:
ztest: Match cleanup ordering between unittest and normal tests

Tests with KERNEL enabled perform their cleanup logic after the suite's
after and test rules are executed. Unittests should do the same.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
Current CI failures appear unrelated to the latest change of moving the unittest cleanup logic

yperess commented:
@nashif any thoughts?

hakehuang commented:
@yperess  which sample / test are you used for validate this change?

yperess commented:
> @yperess which sample / test are you used for validate this change? beside can you rebase those commits into one? as they modify the same file

1. I used tests/ztest/base to add a new test suite that has asserts, assumes, and calls to pass in after/teardown. I cannot check that in because they'll fail twister. We would need to add a feature to twister that will expect the failure and mark it as passing. This seems like too much.
2. I was asked to keep the steps separate in different commits. You can get a single top level view at https://github.com/zephyrproject-rtos/zephyr/pull/48846/files but other people find it easier to review individual commits. It's possible to simply do a squash merge when the PR merges (which I believe is how all PRs should be done). We can ask @fabiobaltieri about that

fabiobaltieri commented:
>     2. I was asked to keep the steps separate in different commits. You can get a single top level view at https://github.com/zephyrproject-rtos/zephyr/pull/48846/files but other people find it easier to review individual commits. It's possible to simply do a squash merge when the PR merges (which I believe is how all PRs should be done). We can ask @fabiobaltieri about that

Yeah the Zephyr guideline is to keep small commits to help reviewability (and every commit should build so that bisect works, https://docs.zephyrproject.org/latest/contribute/guidelines.html#other-commit-expectations). So yeah please don't squash the commits. :-)

Commit:
ztest: add framework tests for failed states

Add specialized tests that execute another test indirectly and ensure
that the results report errors correctly. See the README.rst file
provided with this commit for details.

Signed-off-by: Yuval Peress <peress@google.com>

asemjonovs commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -356,13 +388,15 @@ static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test
 		}
 		run_test_rules(/*is_before=*/false, test, data);
 	}
+	ret |= cleanup_test(test);
```
Add `phase = TEST_PHASE_FRAMEWORK` before `cleanup_test`

yperess replied:
Done

asemjonovs commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -356,13 +388,15 @@ static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test
 		if (suite->after != NULL) {
 			suite->after(data);
 		}
 		run_test_rules(/*is_before=*/false, test, data);
```
Do we still want to run `rules` if anything in `after` asserts?

yperess replied:
I don't think so, an assert in `after` means the binary will be marked as failed. So you won't get any use from running any more test suites until you fix that issue.

asemjonovs commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -356,13 +388,15 @@ static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test
-	if (test_result != ZTEST_RESULT_SUITE_FAIL) {
+	phase = TEST_PHASE_AFTER;
+	if (test_result != ZTEST_RESULT_SUITE_FAIL && !executed_after_fn) {
+		executed_after_fn = true;
```
What you have looks fine, but I have mixed feelings about mixing `goto` and `if(bool)` checks to skip over behavior in the same function.  Could we either create a `out_after` label or possibly clean-up the `setjmp`s and have one jump point, something like 
```
z_test_[pass|skip|fail]() 
{
  longjmp(test_asserted, TC_[PASS|SKIP|FAIL]); // change TC_PASS so != 0
}
:
int run_test() 
{
  ret = setjmp(test_asserted); // Returns 0 in original code. returns TC_[PASS|SKIP|FAIL] from longjmp
  if(ret == 0) { 
    rules(before);
    before(); 
    test_func();
    ret = TC_PASS;
  }
  if(phase != AFTER) {
    phase = AFTER;
    after();
    rules(after);
  }
}
```

yperess replied:
Turns out this wasn't even needed. It was actually caused by a bug I caught later in my code that set the jump to `test_fail` instead of `stack_fail`.

semihalf-barnas-michal commented on tests/ztest/fail/src/main.cpp:
```diff
@@ -0,0 +1,55 @@
+			 * we can prune it
+			 */
+			auto prune_length = result.length() > sizeof(CONFIG_TEST_ERROR_STRING);
+			if (prune_length > 0) {
```
This will be boolean, so you will remove no more than 1 character. Also it should be compared to 2*sizeof, in case if there is a half of the string in the buffer

yperess replied:
Nice catch, thanks!

semihalf-barnas-michal commented on tests/ztest/fail/src/main.cpp:
```diff
@@ -0,0 +1,55 @@
+			}
+		}
+	}
+	auto rc = pclose(pipe) >> 8;
```
Isn't it discarding the return code and reading the id of received signal?

yperess commented:
@nashif would you be able to take a look at this CL? I have another that I think will make others very happy, but it builds on top of this one. Effectively it lets us fail the test run (at the end) if an assume failed. This will make failed assumptions more visible.

jeremybettis commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -311,19 +311,39 @@ void ztest_test_fail(void)
+	case TEST_PHASE_TEARDOWN:
 	case TEST_PHASE_FRAMEWORK:
-		PRINT("\n");
+		PRINT(" ERROR: cannot fail in test '%s()', bailing\n",
```
I think you mean "phase", not "test" here.

yperess replied:
> I think you mean "phase", not "test" here.

Yeah, I might have meant "test phase".... Would you be ok if I fix it in another PR? We have all the approvals for this one and I don't want to reset the review over my poor grammar.

jeremybettis commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -311,19 +311,39 @@ void ztest_test_fail(void)
+	if (phase == TEST_PHASE_TEST) {
+		longjmp(test_pass, 1);
+	}
+	PRINT(" ERROR: cannot pass in test '%s()', bailing\n", get_friendly_phase_name(phase));
```
Likewise.

yperess replied:
Done in #49640

jeremybettis commented on subsys/testsuite/ztest/src/ztest_new.c:
```diff
@@ -311,19 +311,39 @@ void ztest_test_fail(void)
+	case TEST_PHASE_TEST:
+		longjmp(test_skip, 1);
+	default:
+		PRINT(" ERROR: cannot skip in test '%s()', bailing\n",
```
Here also.

yperess replied:
Done in #49640
